### ISSUE TITLE
[Web] BiRefNet_T not working on webgpu

### ISSUE THREAD

**Issue by guschmue** [microsoft/onnxruntime#21968]
### Describe the issue

https://huggingface.co/onnx-community/BiRefNet_T does not work on webgpu
![image](https://github.com/user-attachments/assets/8976c4d3-a599-4a1e-98b1-545e7c38b7db)

### To reproduce

See https://huggingface.co/onnx-community/BiRefNet_T 

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

1.19

### Execution Provider

'webgpu' (WebGPU)

**Comment by guschmue** [comment]
I looked briefly at this: the model takes a lot of memory for activations and will not work with wasm32.
If I run it on webgpu I see it using close to 7GB gpu memory. In theory if your gpu has that kind of memory it should work BUT the model uses GatherND and ScatterND which we have not implemented for webgpu and falls back to wasm just to run out of memory on the wasm end.

**Comment by tidus2102** [comment]
hi @guschmue, thanks for your support. It's great if you can make it works on webgpu (at least with high-end gpus). About the memory, you means we can only run this model with wasm64 right?

**Comment by fs-eire** [comment]
If we can make GatherND and ScatterND working on WebGPU, by using external data side-loading it is possible to make it work in wasm32. And there is an ongoing effort to support wasm64 build.

**Comment by xenova** [comment]
Possibly related: https://huggingface.co/onnx-community/DepthPro-ONNX also throws an error for me:
> An uncaught WebGPU validation error was raised: The number of storage buffers (36) in the Compute stage exceeds the maximum per-stage limit (8).
>  - While validating binding counts
>  - While validating [BindGroupLayoutDescriptor]
>  - While calling [Device].CreateComputePipeline([ComputePipelineDescriptor "Concat"]).

**Comment by xenova** [comment]
https://huggingface.co/briaai/RMBG-2.0 is a new birefnet-based model for state-of-the-art background removal. Would be useful to test too!

**Comment by tidus2102** [comment]
Hi, just a kindly check if there have been any updates or progress regarding this issue? Thank you

**Comment by tidus2102** [comment]
> If we can make GatherND and ScatterND working on WebGPU, by using external data side-loading it is possible to make it work in wasm32. And there is an ongoing effort to support wasm64 build.

hi, how about this?

**Comment by fs-eire** [comment]
#22874 and #22755 introduced implementation of GatherND and ScatterND. Please allow one day or 2 for the pipeline to publish a nightly package.

**Comment by tidus2102** [comment]
Hi, I just test latest (nightly dev build) ORT but still got the memory error when inference BiRefNet_lite onnx model.
![Image](https://github.com/user-attachments/assets/71e8db3e-5879-440e-9094-deb7b64d7a43)
Here is the [sample code](https://codesandbox.io/p/sandbox/8ygzk9).

Please help to check again. Thank you!

**Comment by xenova** [comment]
> the model uses GatherND and ScatterND which we have not implemented for webgpu and falls back to wasm just to run out of memory on the wasm end.
We could implement GatherND and ScatterND what should make it work on high end gpu's but most won't have those high end gpu's.

Now that we have [GatherND](https://github.com/microsoft/onnxruntime/blob/daf9565d1b550fd0d8149e8314bfce61d8df1d54/js/web/lib/wasm/jsep/webgpu/ops/gather-nd.ts) and [ScatterND](https://github.com/microsoft/onnxruntime/blob/daf9565d1b550fd0d8149e8314bfce61d8df1d54/js/web/lib/wasm/jsep/webgpu/ops/scatter-nd.ts) ops implemented, should this be working?

**Comment by xenova** [comment]
I've released a few more birefnet models we can use for testing: https://huggingface.co/models?library=transformers.js&other=birefnet. Unfortunately, it's still an issue today 😅

**Comment by kingback** [comment]
Still no progress?

**Comment by fs-eire** [comment]
The model from https://huggingface.co/onnx-community/BiRefNet_T contains multiple `Split` operator nodes with 32 outputs and a `Concat` operator node with 1024 inputs. This exceeds the hard limit of max buffer bindings in a webgpu shader.

**Comment by fs-eire** [comment]
Technically, the WebGPU EP can bypass the limits by launching the shader multiple times. However, this approach is very inefficient.

If possible, we should make some offline optimizations to the model, to at least avoid generating those operators with too many inputs/outputs.

**Comment by sevagh** [comment]
As an example of @fs-eire suggestion above, I used a script to replace large Concat operators with smaller ones to avoid this webgpu limit.

In my case it was less than 1024, just Inputs=8 Outputs=1, which still exceeds the WebGPU buffer limit of 8 and gave me this error:
>Non-zero status code returned while running Concat node.
 Name:'model_1/harmonic-stacking-diff/concat;model_1/harmonic-stacking-diff/concat'
 Status Message: The number of storage buffer (input=8, output=1) exceeds the limit (8) of the device.

https://gist.github.com/sevagh/23209e01b906bf25cccb4532fb3fd81f 

I don't know how correct it is, ChatGPT helped me write this script, but the inference results seem correct to me (https://github.com/spotify/basic-pitch)

**Comment by fs-eire** [comment]
You need at most 8 (inputs + outputs) according to the error message but you got 9(8+1)

**Comment by sevagh** [comment]
Yeah, I understand it. Just wanted to share a workaround by manipulating the ONNX model.

**Comment by guschmue** [comment]
this one is stale and hopefully no longer relevant.

**Comment by MarcinWolnyEU** [comment]
@guschmue the issue still persists. Tested onnxruntime-node 1.26.0, onnx-proto 8.0.1 .

The script that [sevagh](https://github.com/sevagh) has provided from chatgpt works as a way to bypass the issue - I took his work and built a version in Typescript - [ONNX WebGPU Cascade Fix](https://codeberg.org/MarcinWolnyEU/onnx-webgpu-cascade-fix) - both let you process the onnx files in a way that allows them to be run with no issues. But the core of the problem still persists and it would be amazing if we could get it fixed, instead of relying on these hacky bypasses.
